pkg/resource/policy/hooks.go

```diff
@@ -0,0 +1,193 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package policy
+
+import (
+	"context"
+
+	ackcompare "github.com/aws-controllers-k8s/runtime/pkg/compare"
+	ackcondition "github.com/aws-controllers-k8s/runtime/pkg/condition"
+	ackrtlog "github.com/aws-controllers-k8s/runtime/pkg/runtime/log"
+	svcsdk "github.com/aws/aws-sdk-go/service/iam"
+	corev1 "k8s.io/api/core/v1"
+
+	svcapitypes "github.com/aws-controllers-k8s/iam-controller/apis/v1alpha1"
+)
+
+func (rm *resourceManager) customUpdatePolicy(
+	ctx context.Context,
+	desired *resource,
+	latest *resource,
+	delta *ackcompare.Delta,
+) (*resource, error) {
+	ko := desired.ko.DeepCopy()
+
+	rm.setStatusDefaults(ko)
+
+	if err := rm.syncTags(ctx, &resource{ko}); err != nil {
+		return nil, err
+	}
+	// There really isn't a status of a policy... it either exists or doesn't.
+	// If we get here, that means the update was successful and the desired
+	// state of the policy matches what we provided...
+	ackcondition.SetSynced(&resource{ko}, corev1.ConditionTrue, nil, nil)
+
+	return &resource{ko}, nil
+}
+
+// syncTags examines the Tags in the supplied Policy and calls the
+// ListPolicyTags, TagPolicy and UntagPolicy APIs to ensure that the set of
+// associated Tags  stays in sync with the Policy.Spec.Tags
+func (rm *resourceManager) syncTags(
+	ctx context.Context,
+	r *resource,
+) (err error) {
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("rm.syncTags")
+	defer exit(err)
+	toAdd := []*svcapitypes.Tag{}
+	toDelete := []*svcapitypes.Tag{}
+
+	existingTags, err := rm.getTags(ctx, r)
+	if err != nil {
+		return err
+	}
+
+	for _, t := range r.ko.Spec.Tags {
+		if !inTags(*t.Key, *t.Value, existingTags) {
+			toAdd = append(toAdd, t)
+		}
+	}
+
+	for _, t := range existingTags {
+		if !inTags(*t.Key, *t.Value, r.ko.Spec.Tags) {
+			toDelete = append(toDelete, t)
+		}
+	}
+
+	if len(toAdd) > 0 {
+		for _, t := range toAdd {
+			rlog.Debug("adding tag to policy", "key", *t.Key, "value", *t.Value)
+		}
+		if err = rm.addTags(ctx, r, toAdd); err != nil {
+			return err
+		}
+	}
+
+	if len(toDelete) > 0 {
+		for _, t := range toDelete {
+			rlog.Debug("removing tag from policy", "key", *t.Key, "value", *t.Value)
+		}
+		if err = rm.removeTags(ctx, r, toDelete); err != nil {
+			return err
+		}
+	}
```

Comment:
@RedbackThomson so I'm wondering if the fact that we are doing the deletes AFTER the adds here is the cause of the test failure. When a Tag's *value* only is changed, the Tag will be in *both* the `toAdd` (the updated Tag k/v) and `toDelete` (the original Tag k/v) collections. I'm wondering if the the `RemoveTags` API call only looks at the Tag *key*? If that is the case, the call to RemoveTag will end up removing the Tag erroneously. An easy way of checking if my hypothesis is correct would be to simply change the order above and do the `removeTags` call before the `addTags` call.

Reply:
I think you're right, yeah. You should be able to use `TagPolicy` to update any tag, just by giving it an existing key with a new value. But for the sake of staying consistent with your hooks on `Role`, I'll just switch the order for now and make it do a delete then add. 

I did test updating values in my manual testing and didn't see an issue, but I think combining an update with a delete might be the condition that is causing this problem here.